pkg/route/secret/manager.go

```diff
@@ -0,0 +1,124 @@
+package secret
+
+import (
+	"fmt"
+	"sync"
+	"time"
+
+	v1 "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/client-go/util/workqueue"
+	"k8s.io/klog/v2"
+)
+
+type Manager struct {
+	monitor SecretMonitor
+	// map of handlerRegistrations
+	registeredHandlers map[string]SecretEventHandlerRegistration
+
+	lock sync.RWMutex
+
+	// monitors are the producer of the resourceChanges queue
+	resourceChanges workqueue.RateLimitingInterface
+
+	// secretHandler cache.ResourceEventHandlerFuncs
+	secretHandler cache.ResourceEventHandler
+}
+
+func NewManager(kubeClient *kubernetes.Clientset, queue workqueue.RateLimitingInterface) *Manager {
+	return &Manager{
+		monitor:            NewSecretMonitor(kubeClient),
+		lock:               sync.RWMutex{},
+		resourceChanges:    queue,
+		registeredHandlers: make(map[string]SecretEventHandlerRegistration),
+
+		// default secret handler
+		secretHandler: nil,
+	}
+}
+
+func (m *Manager) WithSecretHandler(handler cache.ResourceEventHandlerFuncs) *Manager {
+	m.secretHandler = handler
+	return m
+}
+
+func (m *Manager) Queue() workqueue.RateLimitingInterface {
+	return m.resourceChanges
+}
+
+func (m *Manager) RegisterRoute(namespace, routeName, secretName string) error {
+	m.lock.Lock()
+	defer m.lock.Unlock()
+
+	// each route (namespace/routeName) should be registered only once with any secret.
+	// Note: inside a namespace multiple different routes can be registered(watch) with a common secret
+	key := generateKey(namespace, routeName)
+	if _, exists := m.registeredHandlers[key]; exists {
+		return apierrors.NewInternalError(fmt.Errorf("route already registered with key %s", key))
+	}
+
+	handlerRegistration, err := m.monitor.AddSecretEventHandler(namespace, secretName, m.secretHandler)
```

Comment:
Probably worth considering how your consuming layers will expect the "route changed secret" to be represented to them.